sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/v2/state/StateDataSourceTestBase.scala

```diff
@@ -18,16 +18,27 @@ package org.apache.spark.sql.execution.datasources.v2.state
 
 import java.sql.Timestamp
 
+import org.scalatest.BeforeAndAfter
+
 import org.apache.spark.sql.{DataFrame, Dataset}
 import org.apache.spark.sql.execution.streaming.MemoryStream
+import org.apache.spark.sql.execution.streaming.state.StateStore
 import org.apache.spark.sql.functions._
 import org.apache.spark.sql.internal.SQLConf
 import org.apache.spark.sql.streaming._
 import org.apache.spark.sql.streaming.util.StreamManualClock
 
-trait StateDataSourceTestBase extends StreamTest with StateStoreMetricsTest {
+trait StateDataSourceTestBase extends StreamTest with BeforeAndAfter with StateStoreMetricsTest {
```

Comment:
`StreamTest` already has `BeforeAndAfterAll`. Can we use `beforeAll` and `afterAll` instead, @chaoqin-li1123 ?

Reply:
The reason we have to clean up StateStore per test is due to maintenance task. When we run the streaming query, state store is being initialized in to the executor, and registration is performed against the coordinator in driver. The lifecycle of the state store provider is not strictly tied to the the lifecycle of the streaming query - the executor closes the state store provider when coordinator indicates to the executor that the state store provider is no longer valid, which is "after" the streaming query has stopped. The lifecycle of the state store provider can overlap among tests.

This means maintenance task against the provider can run after test A. We are clearing the temp directory in test A after the test A has completed, which can break the operation being performed against state store provider being used in test A. E.g. directory does not exist while maintenance task is running.

This won't be an issue in practice because we do not expect the checkpoint location to be temporary, but it is indeed an issue for how we setup and cleanup env for tests.

Reply:
Thank you. Got it. Then, shall we use `beforeEach` and `afterEach` because `SparkFunSuite` has `BeforeAndAfterEach`?

Reply:
This follows the pattern we used in the existing suite.

https://github.com/apache/spark/blob/master/sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingJoinSuite.scala#L42-L54

But I'm fine either way. Good to know we already extend BeforeAndAfterEach by default and no need to extend others.

Reply:
The reason why I requested is that `BeforeAndAfter` is deprecated, @HeartSaVioR . I hope we follow the `scalatest` recommendation in Apache Spark 4.0.0.

- https://www.scalatest.org/scaladoc/1.4.1/org/scalatest/BeforeAndAfter.html
> This trait has been deprecated and will be removed in a future version of ScalaTest. If you are only using its beforeEach and/or afterEach methods, mix in BeforeAndAfterEach instead.

Reply:
Ah, great point. Thanks for the context! @chaoqin-li1123 Shall we update the trait?